### PR TITLE
User Timing L3 changes

### PR DESCRIPTION
**Pull request by npm1**
See https://github.com/w3c/user-timing/issues/43

***
<a href="https://pr-preview.s3.amazonaws.com/npm1/user-timing/pull/46.html" title="Last updated on Dec 7, 2018, 7:04 PM GMT (966bb10)">Preview</a> | <a href="https://pr-preview.s3.amazonaws.com/w3c/user-timing/46/c726dad...npm1:966bb10.html" title="Last updated on Dec 7, 2018, 7:04 PM GMT (966bb10)">Diff</a>

### CODE REVIEW

**Comment by igrigorik** [comment]
Paging sir @philipwalton for a review :-)

**Comment by npm1** [comment]
Hold off review until I update the PR with the simplified proposal for the API shape. I'll message when I'm done.

**Comment by npm1** [comment]
Ok, I think this is ready for review :)

File: index.html [hunk 1]
```diff
@@ -124,27 +147,62 @@ <h2><dfn>User Timing</dfn></h2>
...
           <li>Create a new <a>PerformanceMark</a> object (<var>entry</var>).</li>
           <li>Set <var>entry</var>'s <code>name</code> attribute to <var>markName</var>.</li>
           <li>Set <var>entry</var>'s <code>entryType</code> attribute to <code>DOMString "mark"</code>.</li>
-          <li>Set <var>entry</var>'s <code>startTime</code> attribute to the value that would be returned by the <code>Performance</code> object's <a data-cite="!HR-TIME-2#dom-performance-now"><code>now()</code></a> method.</li>
+          <li>
+            Set <var>entry</var>'s <code>startTime</code> attribute as follows:
+            <ol>
+              <li>If <var>markOptions</var> is present and it's <a>startTime</a> member is present, then set it to the value of <var>markOptions</var>'s <a>startTime</a>.</li>
```
[the remark this answers is not in the document]
Comment (npm1) [reply index.html]: Done

File: index.html [hunk 2]
```diff
@@ -155,32 +213,75 @@ <h2><dfn>clearMarks()</dfn> method</h2>
...
         </ol>
       </section>
-      <section>
+      <section data-link-for="PerformanceMeasureOptions">
         <h2><dfn>measure()</dfn> method</h2>
         <p>Stores the <code>DOMHighResTimeStamp</code> duration between two marks along with the associated name (a "measure"). It MUST run these steps:</p>
         <ol>
-          <li>Let <var>end time</var> be <code>0</code>.</li>
-          <li>If <var>endMark</var> is omitted, let <var>end time</var> be the value that would be returned by the <code>Performance</code> object's <a data-cite="!HR-TIME-2#dom-performance-now"><code>now()</code></a> method. Otherwise:
+          <li>If <var>startOrMeasureOptions</var> is present and a <a>PerformanceMeasureOptions</a> object, and if <var>endMark</var> is present, <a data-cite="!WEBIDL#dfn-throw">throw</a> a <a data-cite="!WEBIDL#syntaxerror"><code>SyntaxError</code></a>.</li>
```
[the remark this answers is not in the document]
Comment (npm1) [reply index.html]: Done

**Comment by igrigorik** [comment]
@npm1 before we land this, let's also make sure we tag this as L3.

- I believe all the L2 work is addressed? If so, let's snapshot L2 as a branch or tag.
- We should update the level in the header to make it clear that this is L3

/cc @yoavweiss @plehegar @toddreifsteck

**Comment by npm1** [comment]
Ah I thought that the snapshot was done already, since https://www.w3.org/TR/user-timing-2/ works and since the 'outstanding issues' PR from Philip was merged on February. I can't do this (don't know how, and wouldn't have the permissions needed), so who would be a reasonable person to ping about creating the snapshot?

Re updating the header, done and also updated the 'Status of This Document' part. I assume the respecConfig is not to be updated yet. @igrigorik @plehegar please take a look at the latest commit to make sure those changes are correct.

**Comment by igrigorik** [comment]
@plehegar @siusin whats your preferred and recommend way to snapshot L2? Would creating a tag be enough so we have something to reference later?

@npm1 LGTM, thanks for the great work on this one! Once we hear back from Philippe and Xiaoqian, this is good to merge.

**Comment by siusin** [comment]
@npm1 you should have received an invitation from GitHub to share the admin access of this repo.

There is 0 open issue, and only 1 failing test for L2 now, so I'd suggest:
- create a new branch for L2, and use the gh-pages branch as the main entry to edit and auto-pub L3;
- if the group agree that the test [mark-measure-return-null.html](https://github.com/web-platform-tests/wpt/commit/a4efb68eb4bb30453a6c958ad58f8c2540dc967e) is only for feature detection, and there is no need to change the L2 spec for this failing test, Plh and I will send a Transition Request to move L2 to Proposed Rec to the W3C members once we get a group consensus (maybe from the next WG call?).

**Comment by igrigorik** [comment]
Thanks @siusin. 

+1 for creating an L2 branch. @npm1 can you push that before we merge this into gh-pages? The rest is non-blocking for this PR. 

---

> * if the group agree that the test [mark-measure-return-null.html](https://github.com/web-platform-tests/wpt/commit/a4efb68eb4bb30453a6c958ad58f8c2540dc967e) is only for feature detection, and there is no need to change the L2 spec for this failing test, Plh and I will send a Transition Request to move L2 to Proposed Rec to the W3C members once we get a group consensus (maybe from the next WG call?).

That test looks reasonable to me, why or where is it failing? /cc @yoavweiss @philipwalton for another pair of eyes.

**Comment by siusin** [comment]
Test results for UT in the experimental version:
https://wpt.fyi/results/user-timing?aligned&product=chrome%5Bexperimental%5D&product=edge%5Bstable%5D&product=firefox%5Bexperimental%5D&product=safari%5Bexperimental%5D

[mark-measure-return-null.html](https://wpt.fyi/results/user-timing/mark-measure-return-null.html?product=chrome%5Bexperimental%5D&product=edge%5Bstable%5D&product=firefox%5Bexperimental%5D&product=safari%5Bexperimental%5D&aligned) only fails in Edge, Safari and Firefox.

**Comment by npm1** [comment]
Trying it out on Firefox console, it seems that if you do `const measure = performance.measure("measure");` then `measure === undefined`. That seems correct to me, whereas in Chrome `measure === null`. Overall this test seems to provide no value to testing the usage of User Timing so I propose we remove this test. Thoughts?

**Comment by npm1** [comment]
Also, I see two options for the L2 branch. One is fc644a04606c190a65948670efd9b680ae41d4e0 to exclude registering entry types which is a pretty new addition. Otherwise we can point it to c726dad5e59889584dbf06d5b1d6ebd8533e133b

**Comment by npm1** [comment]
Ah and I guess the other option is for the test, if we want to keep it, is to just accept null or undefined. That should make it pass in all implementations, I think?

**Comment by igrigorik** [comment]
@npm1 I was puzzling over the same test failures with undefined vs null.. it seems odd to me that we have different implementations returning different things here? That aside, the test is verifying return behavior that's updated in L3, so it seems reasonable to keep it?

Re, branch point: we should freeze it at same point as when we published the CR.

Paging @yoavweiss @toddreifsteck for input on the test.

**Comment by npm1** [comment]
@igrigorik I believe the reason we're returning null is because we have L3 under a flag which means the IDL is not quite in the L2 format. Given that, if we want to keep the tests it may also make sense to change null to undefined in the tests and hence Chrome would be failing them. The tests won't make sense when L3 ships since the methods should no longer have an undefined return value. They would need to be modified appropriately. Overall I don't think this test should be blocking L2.

Re branch point: not sure exactly what you mean. Is it f81d84fd7ae3017ab193a8c2521387562ce7f4ee? specStatus was changed to CR in that commit.

**Comment by igrigorik** [comment]
> The tests won't make sense when L3 ships since the methods should no longer have an undefined return value. They would need to be modified appropriately.

Ah, that's a great point. With that in mind, yeah I agree that we should consider dropping it.

**Comment by siusin** [comment]
@npm1 sounds good! I'll send a request to move L2 to Proposed REC once the test is landed.

**Comment by npm1** [comment]
@siusin the modified test has landed[1]. Let me know when it is fine to merge this PR. Thanks!

[1] https://github.com/web-platform-tests/wpt/pull/14511

**Comment by siusin** [comment]
TransReq sent:
https://lists.w3.org/Archives/Member/chairs/2018OctDec/0097.html

Please merge the `gh-pages` changes to the current `l2` branch before landing the L3 features. We'll publish a First Public Working Draft for L3 once this PR lands.

**Comment by npm1** [comment]
@siusin l2 branch already points to the desired commit. The current difference between l2 and gh-pages consists of commits which we do not want in l2. So I'm merging this PR.

**Pull request merged**